telinit.man
.TH {telinit} 8
'''
.SH NAME
{telinit} \- pass commands to a running \fB{init}\fR(8) process.
'''
.SH SYNOPSIS
\fB{telinit}\fR \fIrunlevel\fR
.br
\fB{telinit}\fR {\fB?\fR|\fBq\fR|\fBstatus\fR|\fBreload\fR}
.br
\fB{telinit}\fR {\fBpoweroff\fR|\fBhalt\fR|\fBreboot\fR|\fBsleep\fR|\fBsuspend\fR}
.br
\fB{telinit}\fR {\fBstart\fR|\fBstop\fR} \fIname\fR [\fIname\fR ... ]
'''
.SH DESCRIPTION
The only way to contact a running init is via @initctl socket.
{telinit} does exactly that, providing also readable shortcuts
for raw init commands.
.P
Whatever output init may produce is passed to {telinit}'s stdout.
'''
.SH COMMANDS
.IP [0123456789]
Initiate a switch to specified runlevel.
.IP [+-][abcdef]
Activate/deactivate specified sublevel.
.IP "\fB?\fR, \fBlist\fR" 4
Request current status and process list from init.
.IP "\fBreload\fR" 4
Reload configuration.
.IP "\fBhalt\fR, \fBreboot\fR, \fBpoweroff\fR" 4
Stop the system, possibly requesting reboot or poweroff.
All three commands request a switch to runlevel 0, followed
by \fBreboot\fR(2) call with LINUX_REBOOT_CMD_HALT, LINUX_REBOOT_CMD_RESTART
or LINUX_REBOOT_CMD_POWER_OFF argument respectively.
.IP "\fBsleep\fR, \fBsuspend\fR" 4
Initiate a switch to runlevels 8 and 9 respectively.
For these commands to work properly, relevant actions must be configured
in inittab; init itself will not put the system to sleep.
.IP "\fBstart\fR \fIname\fR" 4
Start entry \fIname\fR. In case the entry is not configured to be started on
current runlevel, adjust its mask settings and start anyway.
.IP "\fBrestart\fR \fIname\fR" 4
Kill named process without changing its runlevel mask, forcing immediate respawn.
.IP "\fBstop\fR \fIname\fR" 4
Stop entry \fIname\fR.
This command marks the entry as disabled without changing runlevel mask.
.IP "\fBunstop\fR \fIname\fR" 4
Undo \fBstop\fR command without changing runlevel mask like \fBstart\fR does.
Depending on entry configuration, it may or may not be started immediately.
.IP "\fBpause\fR \fIname\fR, \fBresume\fR \fIname\fR" 4
Send SIGSTOP or SIGCONT to the process.
.IP "\fBhup \fIname\fR\fR" 4
Send SIGHUP to the process.
.IP "\fBpidof \fIname\fR\fR" 4
Print process id for the named entry.
.P
Wherever applicable, \fIname\fR is the first field from \fB{inittab}\fR(5) line,
or file basename in \fBinitdir\fR(5).
'''
.SH FILES
.IP {@initctl}
\fBunix\fR(7) domain socket {telinit} uses to contact init
'''
.SH CAVEATS
Runlevel change and process status change commands sent by {telinit}
are only executed after {telinit} has terminated the connection.
The only errors reported for these commands are those apparent immediately
by examining process record. Something like "named entry not found" will
be reported, but \fBexecve\fR(2) failure won't, at least not by {telinit}
(syslog will probably have it).
.P
Runlevel switch requests only start the move to another runlevel.
If some of the commands take long enough, it maybe be possible to see
init in-between runlevels with {telinit} q, and even change the target runlevel
before the switch is completed.
.P
On the other hand, {telinit} does wait for reconfiguration to finish before
exiting, and will show all configuration errors.
'''
.SH NOTES
This page describes {telinit} from \fB{INIT}\fR package.
Other init implementations may behave differently.
.P
In particular, this version of {telinit} is not compatible with sysvinit.
Most sysvinit options are not supported. Runlevel change and configuration
reload commands are same in both implementations however.
.P
Runlevel 0 is treated as a special case by {init}, prompting a reboot(2) call.
This is unlike sysvinit, for which "0 means halt" is merely a configuration
convention. On the other hand, runlevel 6 is not a special runlevel for {init},
and "{telinit} 6" will not reboot the system unless configuration explicitly
prompts so. Use "{telinit} reboot" instead.
'''
.SH SEE ALSO
\fB{init}\fR(8), \fB{inittab}\fR(5)


telinit.c
#define	_GNU_SOURCE
#include <unistd.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <string.h>
#include <errno.h>

#include "config.h"

#define ERR ((void*) -1)

/* Telinit sends commands to init via its control socket and reads init's
   output back. One command is sent per connection. In case there are
   multiple arguments, telinit sends one at a time and reconnects between
   sending them. There is no multiplexing of any sort, it's write-all
   followed by read-all for each command. */

static int runcmd(const char* cmd);
static void die(const char* msg, const char* arg);

/* For convenience, one-letter init command codes (cc's) are given readable
   names within telinit. These should be kept in sync with init_cmds.c.
   Sleep commands are "telinit-only", they send generic level switch cc's. */

static struct cmdrec {
	char cc;
	char arg;
	char name[10];
} cmdtbl[] = {
	/* halt */
	{ 'H', 0, "halt",	},
	{ 'P', 0, "poweroff",	},
	{ 'R', 0, "reboot",	},
	/* process ops */
	{ 'p', 1, "pause",	},
	{ 'w', 1, "resume",	},
	{ 'h', 1, "hup",	},
	{ 's', 1, "start",	},
	{ 'r', 1, "restart",	},
	{ 't', 1, "stop",	},
	{ 'u', 1, "unstop",	},
	/* state query */
	{ '?', 0, "list",	},
	{ 'i', 1, "pidof",	},
	/* reconfigure */
	{ 'c', 0, "reload",	},
	/* slippery levels */
	{ '7', 0, "doze",	},
	{ '8', 0, "sleep",	},
	{ '9', 0, "suspend",	},
	{  0  }
};

int main(int argc, char** argv)
{
	struct cmdrec* cr = NULL;
	char buf[NAMELEN+2];

	int hasarg = 0;
	char* ptr = buf;
	char* cmd = argv[1];
	char* cm1 = cmd + 1;
	int i;

	if(argc < 2)
		die("Usage: telinit cmd [args]", NULL);

	if((*cmd >= '0' && *cmd <= '9') || *cmd == '-' || *cmd == '+') {
		ptr = cmd;
	} else {
		for(cr = cmdtbl; cr->cc; cr++) {
			if(!*cm1 && *cmd == cr->cc)
				break;
			if(!strcmp(cmd, cr->name))
				break;
		} if(!cr->cc)
			die("Unknown command ", cmd);

		buf[0] = cr->cc;
		buf[1] = '\0';
		hasarg = cr->arg;
	}

	int ret = 0;

	if(!hasarg)
		ret = runcmd(ptr);
	else for(i = 2; i < argc; i++) {
		memset(buf + 1, 0, sizeof(buf)-1);
		strncpy(buf + 1, argv[i], sizeof(buf)-2);
		ret |= runcmd(buf);
	}

	return (ret ? 1 : 0);
}

static int opensocket(void)
{
	int fd;
	struct sockaddr_un addr = {
		.sun_family = AF_UNIX,
		.sun_path = INITCTL
	};

	if(addr.sun_path[0] == '@')
		addr.sun_path[0] = '\0';

	fd = socket(AF_UNIX, SOCK_STREAM, 0);	
	if(fd < 0)
		die("Can't create socket: ", ERR);
	if(connect(fd, (struct sockaddr*)&addr, sizeof(addr)))
		die("Can't connect to " INITCTL ": ", ERR);

	return fd;
}

/* Init can only be controlled by root. This is enforced by sending
   user credentials in auxiliary data. Telinit could have checked it
   too, giving early non-root access error, but it's not done. First,
   the code is already in init, and second, there DEVMODE. */

static int sendcmd(int fd, const char* cmd)
{
	int len = strlen(cmd);

	if(write(fd, cmd, len) < 0)
		die("write failed: ", ERR);
	
	return 0;
}

/* The tricky part here is demuxing init output, which can be error
   messages to be sent to stderr, or pidof/list output which is stdout
   kind of data.

   Init has a very specific reply pattern, it's eiter all-error
   or all-non-error, so a simple # indicator at the start of init
   output is used to choose the fd.

   As a side effect, this also determines telinit return code.
   Non-empty error message means there was an error, which empty
   output or any #-output means everything went well. */

static int recvreply(int fd)
{
	char buf[100];
	int rr;
	int out = 0;
	int ret = 0;
	int off = 0;

	while((rr = read(fd, buf, sizeof(buf))) > 0) {
		if(!out) {
			if(buf[0] == '#') {
				off = 1;
				out = 1;
			} else {
				ret = -1;
				out = 2;
			}
		} else if(off) off = 0;

		write(out, buf + off, rr - off);
	}

	return ret;
}

static int runcmd(const char* cmd)
{
	int fd;
	int r = 0;

	fd = opensocket();
	sendcmd(fd, cmd);
	shutdown(fd, SHUT_WR);
	r = recvreply(fd);
	close(fd);

	return r;
};

static void die(const char* msg, const char* arg)
{
	char buf[256];
	int len = strlen(msg);
	int max = sizeof(buf) - 2;

	strncpy(buf, msg, max);

	if(arg == ERR)
		arg = strerror(errno);
	if(arg)
		strncpy(buf + len, arg, max - len);

	len = strlen(buf);
	buf[len++] = '\n';

	write(2, buf, len);
	_exit(-1);
}
